src/transfer_zip.py

```diff
@@ -0,0 +1,657 @@
+
+    return ns
+
+
+def transfer_data_query(data_query: DataQuery) -> None:
+    """Transfer all files matching a data query into zip files.
+
+    Parameters
+    ----------
+    data_query: `DataQuery`
+        The query and associated embargo time.
+    """
+    global logger, config, source_butler
+
+    # End of window is now - embargo length
+    end_time = config.now - TimeDelta(data_query.embargo_hours * 3600, format="sec")
+    # If window is defined, then start is that much before the end
+    # Otherwise, start is infinitely previous
+    if config.window is not None:
+        start_time = end_time - TimeDelta(config.window, format="quantity_str")
+    else:
+        start_time = Time(0, format="jd")
+    ok_timespan = Timespan(start_time, end_time)
+
+    # Find all exposures meeting criteria
+    dim_where = "(exposure.timespan OVERLAPS _ok_timespan)"
+    dim_where += f" AND ({data_query.where})" if data_query.where else ""
+    dim_bind = {"_ok_timespan": ok_timespan}
+    logger.info("Querying exposure: %s with %s", dim_where, dim_bind)
+    try:
+        exposures = source_butler.query_dimension_records(
+            "exposure",
+            where=dim_where,
+            bind=dim_bind,
+            limit=None,
+            instrument=data_query.instrument,
+        )
+    except EmptyQueryResultError:
+        logger.info("No matching records")
+        return
+    logger.info(
+        "Result is %d exposures from %s to %s",
+        len(exposures),
+        exposures[0].id,
+        exposures[-1].id,
+    )
+
+    for exp in exposures:
+        process_exposure(exp, data_query.instrument)
+
+
+def process_exposure(exp: DimensionRecord, instrument: str) -> None:
+    """Process an exposure by zipping, ingesting, and registering it in Rucio.
+
+    Parameters
+    ----------
+    exp: `lsst.daf.butler.DimensionRecord`
+        The exposure to process.
+    instrument: `str`
+        The name of the instrument corresponding to the exposure.
+    """
+    global logger, config, source_butler, dest_butler, rucio_interface
+
+    # Check several times for existence of the result to avoid work
+    # in case of race conditions
+    zip_name = f"{exp.obs_id}.zip"
+    dest_dir = (
+        ResourcePath(config.dest_uri_prefix).join(instrument).join(f"{exp.day_obs}")
+    )
+    dest_path = dest_dir.join(zip_name)
+    if dest_path.exists():
+        logger.info("Zip exists, skipping processing: %s", dest_path)
+        return
+
+    # Map exposure to tracts
+    tracts = set()
+    with source_butler.query() as q:
+        q = q.join_dimensions(["tract"]).where(
+            "visit = _exposure",
+            instrument=instrument,
+            skymap="lsst_cells_v1",
+            bind={"_exposure": exp.id},
+        )
+        for id in q.data_ids(["tract"]):
+            tracts.add(id["tract"])
+
+    # Find all datasets for this exposure and its source directory
+    refs = source_butler.query_datasets(
```

Comment:
You are assuming that it's not possible for an exposure record to exist without a dataset existing? (you aren't setting `explain=False`). When we purge old raws are we purging dimension records?  I assume you are not doing this as a global `query_datasets` up front because you want to limit the number of refs in memory.

Reply:
Even if we leave exposure records, the destination zip file should exist.  But yes, if possible, I think it would be nice to purge dimension records from the database when we purge all relevant datasets from Embargo.  A `query_datasets` for even a night is hundreds of thousands of refs; I figured a few queries per exposure (not even necessarily going to the database) was better than grouping the results from a single huge query.